### ISSUE TITLE
Recursive function calls & Functions that are Parameters

### ISSUE THREAD

**Issue by akalicki** [odds-lang/odds#66]
Open thought question: how do we handle recursive function calls? Functions require evaluating their bodies and return expressions in order to infer the `Func` type, but where they are called recursively not everything has been evaluated so checking against constraints throws errors.

Possibility: check all non-recursive statements in function, infer type, then check recursive statements?

**Comment by dannyech** [comment]
Ok i think i figured a way to do this: basically in the env have another variable called something like `cur_function` that points to the `ss_id` and `Sast.func` of the function that we are currently in. Then, there are two cases:

1. the function is referenced recursively __before__ any of its parameters have been constrained. If this is the case, we constrain the function params and return on the recursive call.

2. The function is referenced recursively __after__ its parameters have been constrained. If this is the case, we check against the `Sast.func`in the current env but still try to constrain the return.

This seem like a good solution?

**Comment by dannyech** [comment]
Also keep in mind that our solution has to allow for mutual recursion within our language. (A parent function must be able to be called (recursively) within a nested child function. 

So we would have to use a map of `ss_id` -> `func` so we could keep a running tally of all the functions that we are nested within, and that way be able to call and constrain them...

**Comment by dannyech** [comment]
Also: we basically already have this map in the env - the difference is now we constrain function params and return types as we move through the statement list.

**Comment by dannyech** [comment]
basically we can constrain functions as we go like any other variable, only when we get to the end of a function, if its ret_type is `Unconst` do we reject it. Going to try to implement this

**Comment by dannyech** [comment]
@akalicki 
Alright, I've been thinking about this awhile. Here is what we need to do:

1. Add a type called `Any`. The type `Any` means, as you might guess, that the parameter/arg/whatever can be any type - it does not matter. For example, `print` would take `Any` and the `==` operator would take `Any` on either side. We need `Any` because right now `Unconst` is doing two jobs: one, to indicate that function might take any param, and two, to indicate that a param needs to be constrained still.

2. Now that we have a new type `Any`, it means that anytime we come across a function call in which either the **return type or the params type** are Unconst (i.e. on a recursive call), we constrain the function like we would any other variable. At the end of check_fdecl, if the return type is still Unconst, we still reject the function like we have been.

For example,
```ocaml
do foo = (x) -> (* foo's type is (Unconst => Unconst) *)
  do i = foo(2) (* foo's type is (Num => Unconst) and i is Unconst*)
  do y = i + 2  (* y is Num and i is constrained to Num *)
return i (* foo's type is now (Num => Num) *)
```

what do you think?

**Comment by akalicki** [comment]
The example looks good to me! How does `Any` play into it?

**Comment by dannyech** [comment]
Checklist of changes that need to be made:

- [ ] Add `Any` type to Sast. Add `Any` to appropriate places in Analyzer.
- [ ] Change `check_fcall` and `check_fcall_args` to allow the constraining of functions
- [ ] Add constraining behavior for functions in `constrain_ew`

**Comment by dannyech** [comment]
@akalicki Right so `Any` means that we won't constrain the function on the call.

If we didn't have `Any` then the first time we come across a `print` we are going to try to constrain it, but we do not want this to happen.

**Comment by akalicki** [comment]
Gotcha that makes sense! This looks good to me then. Do you want to do it as part of the same inference PR I have out or as a separate add on?

**Comment by dannyech** [comment]
i made another branch and have been trying things for a while. It's really disorganized tho. I think i might delete the branch. So if we merge your PR, i'll just branch off of Master Refactor after we've done that. That way I start on a clean slate and can follow my checklist.

**Comment by dannyech** [comment]
@akalicki Should I go ahead and merge you're PR?

**Comment by akalicki** [comment]
I'm cool with it! Seeing as we've had 2 eyes look over pretty much everything already and it's just going into the master refactor branch anyway

**Comment by dannyech** [comment]
Did it!

**Comment by dannyech** [comment]
Ok wait

**Comment by dannyech** [comment]
@akalicki Ok wait this might be an issue:

say we have:
```ocaml
do foo = (x) ->
  do foo(x)
  do foo(2)
  return x
```
should we be able to constrain `x`? Obviously `x` has to be a Num, but it's coming before we get to the part where we can constrain `foo`'s params to be a Num, so the above function is illegal in our language. Do you see any problems with this? Obviously no one would ever write a function like this, but is their a case when you would blindly feed a param to a recursive call without doing anything to the param at all?

```ocaml
do foo = (x) ->
  do foo(x + 2)
  do foo(2)
  return x
```

^ this function would be valid bc of the `x + 2`

**Comment by dannyech** [comment]
note to self: 

- [ ] constrain function param variables in addition to function type itself

**Comment by dannyech** [comment]
New Issue: 

We need relate param vars to their corresponding param_type in Func. If we do this, any time we constrain a param, we make sure to constrain the Func type, and anytime we constrain the Func type, we make sure to constrain the param

**Comment by dannyech** [comment]
#### New issue:

we must be able to constrain the type of params of functions that are paremeters, and ensure that any types marked `Unconst` at the end of the function become `Any`

 - [ ] ensure that when we compare against `Unconst` in check_type_mismatch, we also check to see if any of the parameters have been constrained if type is `Func`, and if not, make them `Any`
   the issue is with these lines of code:
  
One: 
```ocaml
   if var.s_type = Unconst && func_param_type = Unconst then 
          check_params_type_mismatch env (Any :: acc) (List.tl func_param_types) tl
 ```

Two: 
```ocaml
else if func_param_type = Unconst then
          check_params_type_mismatch env (var.s_type :: acc) (List.tl func_param_types) tl
```

Three: 
```ocaml
else if var.s_type = Unconst then
          let env' = constrain_e func_env (Sast.Id(ssid)) func_param_type in 
          check_params_type_mismatch env' (func_param_type :: acc) (List.tl func_param_types) tl
```

TL;DR the raw compare `=` is not good enough, we need a more complex compare to see if - if the type is a Func - if any of the params have been constrained in either the var or the param

**Comment by dannyech** [comment]
notes to self:
  ###### fix:
- [ ] `f(2) + 2` constrain's `f`'s return type but not its arg type if `f` is a parameter. why?????
- [ ] Some funky stuff will happen if you feed collect_constraints an any function - see if there will be issues and resolve if possible...

**Comment by akalicki** [comment]
@dannyech anything more that needs to be done on this issue? or good to close?

**Comment by dannyech** [comment]
@akalicki So i think i solved the final issue we were having on the new branch i made (constrain_any) - try to break it if you can. If not, I'll make a pull request, we can merge, and then close the issue.

**Comment by akalicki** [comment]
Looks good to me. Added a couple comments on the commits for when you PR. Mind adding a couple pass tests for things that should now work and fail tests for things that should not?
